docs/general/cairo_precompiles.md

```diff
@@ -0,0 +1,225 @@
+    uint256 constant FUNCTION_SELECTOR_SET_COUNTER = uint256(keccak256("set_counter")) % 2**250;
+
+    /// @dev The cairo function selector to call - `get`
+    uint256 constant FUNCTION_SELECTOR_GET = uint256(keccak256("get")) % 2**250;
+
+
+    constructor(uint256 cairoContractAddress) {
+        cairoCounterAddress = cairoContractAddress;
+    }
+
+    function getCairoCounter() public view returns (uint256 counterValue) {
+        // `get_counter` takes no arguments, so data is empty
+        uint256[] memory data;
+        bytes memory returnData = CairoLib.staticcallContract(cairoCounterAddress, FUNCTION_SELECTOR_GET, data);
+
+        // The return data is a 256-bit integer, so we can directly cast it to uint256
+        return abi.decode(returnData, (uint256));
+    }
+
+    /// @notice Calls the Cairo contract to increment its internal counter
+    function incrementCairoCounter() external {
+        // `inc` takes no arguments, so data is empty
+        uint256[] memory data;
+        CairoLib.callContract(cairoCounterAddress, FUNCTION_SELECTOR_INC, data);
+    }
+
+    /// @notice Calls the Cairo contract to set its internal counter to an arbitrary value
+    /// @dev The counter value is split into two 128-bit values to match the Cairo contract's expected inputs (u256 is composed of two u128s)
+    /// @param newCounter The new counter value to set
+    function setCairoCounter(uint256 newCounter) external{
+        // The u256 input must be split into two u128 values to match the expected cairo input
+        uint128 newCounterLow = uint128(newCounter);
+        uint128 newCounterHigh = uint128(newCounter >> 128);
+
+        uint256[] memory data = new uint256[](2);
+        data[0] = uint256(newCounterLow);
+        data[1] = uint256(newCounterHigh);
+        CairoLib.callContract(cairoCounterAddress, FUNCTION_SELECTOR_SET_COUNTER, data);
+    }
+}
+
+```
+
+Once deployed, the contract can be called to increment the counter in a Cairo
+contract deployed at starknet address `cairoCounterAddress`. The deployment
+address will need to be communicated to Kakarot for the precompile to be
+whitelisted.
+
+Internally, a new logic flow is implemented when processing message calls. If
+the target address is the Cairo precompile address, we check if the code_address
+of the message is whitelisted. If it is, we execute the Cairo contract. If it is
+not, we revert the transaction.
+
+To execute the Cairo contract, we need to convert the EVM calldata, expressed as
+a `bytes[]`, to the expected Cairo inputs, serialized to an `Array<felt252>`.
+Each 256-bit word sequence in the EVM calldata must correspond to an element of
+at most, 251 bit, Cairo's native field element. For example, consider the
+`setCairoCounter` function above. The `newCounter` argument is 256 bits, and
+thus split into two 128-bit values to match the expected Cairo input. If we want
+to increase the counter by 1, the data sent would be `data = ["00" * 63 + "01"].
```

Comment:
I'd just say that cairo calldata is a `felt252[]` so that the `data` in solidity is a `uint256[]` and each `uint256` will be cast to a `felt`